src/Agent.Worker/Variables.cs

```diff
@@ -90,6 +149,135 @@ public bool TryGetValue(string name, out string val)
             _trace.Verbose($"Get '{name}' (not found)");
             return false;
         }
+
+        private void RecursivelyExpand(out List<string> warnings)
```

Comment:
I tried to make this algorithm as simple as possible. I refactored the code over and over again, left extensive comments, and even refined the comments over and over again.